crates/bitwarden/src/tool/generators/password.rs

```diff
@@ -40,10 +41,249 @@ pub struct PassphraseGeneratorRequest {
+    password_with_rng(rand::thread_rng(), input)
+}
+
+pub(super) fn password_with_rng(
+    mut rng: impl RngCore,
+    input: PasswordGeneratorRequest,
+) -> Result<String> {
+    // We always have to have at least one character set enabled
+    if !input.lowercase && !input.uppercase && !input.numbers && !input.special {
+        return Err(Error::Internal(
+            "At least one character set must be enabled",
+        ));
+    }
+
+    // Generate all character dictionaries
+    let chars = PasswordGeneratorCharSet::new(
+        input.lowercase,
+        input.uppercase,
+        input.numbers,
+        input.special,
+        input.avoid_ambiguous.unwrap_or(false),
+    );
+
+    // Make sure the minimum values are zero when the character
+    // set is disabled, and at least one when it's enabled
+    fn get_minimum(min: Option<u8>, enabled: bool) -> u8 {
+        if enabled {
+            u8::max(min.unwrap_or(1), 1)
+        } else {
+            0
+        }
+    }
+    let min_lowercase = get_minimum(input.min_lowercase, input.lowercase);
+    let min_uppercase = get_minimum(input.min_uppercase, input.uppercase);
+    let min_number = get_minimum(input.min_number, input.numbers);
+    let min_special = get_minimum(input.min_special, input.special);
+
+    // Check that the minimum lengths aren't larger than the password length
+    let min_length = min_lowercase + min_uppercase + min_number + min_special;
+    let length = input.length.unwrap_or(DEFAULT_PASSWORD_LENGTH);
+    if min_length > length {
+        return Err(Error::Internal(
+            "Password length can't be less than the sum of the minimums",
+        ));
+    }
+
+    // Generate the minimum chars of each type, then generate the rest to fill the expected length
+    let mut buf = Vec::with_capacity(length as usize);
+
+    for _ in 0..min_lowercase {
+        buf.push(*chars.lower.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in 0..min_uppercase {
+        buf.push(*chars.upper.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in 0..min_number {
+        buf.push(*chars.number.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in 0..min_special {
+        buf.push(*chars.special.choose(&mut rng).expect("slice is not empty"));
+    }
+    for _ in min_length..length {
+        buf.push(*chars.all.choose(&mut rng).expect("slice is not empty"));
+    }
+
+    buf.shuffle(&mut rng);
+    Ok(buf.iter().collect())
 }
 
 pub(super) fn passphrase(_input: PassphraseGeneratorRequest) -> Result<String> {
     Ok("correct-horse-battery-staple".to_string())
 }
+
+#[cfg(test)]
+mod test {
+    use rand::SeedableRng;
+
+    use super::*;
+
+    // We convert the slices to Strings to be able to use `contains`
+    // This wouldn't work if the character sets were ordered differently, but that's not the case for us
+    fn to_string(chars: &[char]) -> String {
+        chars.iter().collect()
+    }
```

Comment:
❓ I wonder how this would look if instead you used [HashSet::difference](https://doc.rust-lang.org/std/collections/hash_set/struct.HashSet.html#method.difference). Main reason is because it gets away from the whole ordering business. I have no clue what would happen with execution time.

Reply:
Good point! I've changed it to use HashSet::is_superset and I think it makes more clear what the goal of the test is, plus now ordering is not an issue.

Tests still seem to run almost instantly, so I don't think it's going to be a problem.